libvast/src/arrow_table_slice.cpp

```diff
@@ -1277,4 +1278,155 @@ template class arrow_table_slice<fbs::table_slice::arrow::v0>;
+                 std::shared_ptr<arrow::DataType> arrow_type) {
+  auto f = detail::overload{
+    [&](const std::shared_ptr<arrow::Array>& a,
+        const std::shared_ptr<arrow::DataType>& at)
+      -> std::shared_ptr<arrow::Array> {
+      VAST_ASSERT(a->type()->Equals(at));
+      return a;
+    },
+    [&](const std::shared_ptr<arrow::StringArray>& pa,
+        const std::shared_ptr<pattern_extension_type>& pt)
+      -> std::shared_ptr<arrow::Array> {
+      return std::make_shared<pattern_array>(pt, pa);
+    },
+    [&](const std::shared_ptr<arrow::FixedSizeBinaryArray>& ba,
+        const std::shared_ptr<address_extension_type>& at)
+      -> std::shared_ptr<arrow::Array> {
+      VAST_ASSERT(ba->byte_width() == 16);
+      return std::make_shared<address_array>(at, ba);
+    },
+    [&](const std::shared_ptr<arrow::FixedSizeBinaryArray>& ba,
+        const std::shared_ptr<subnet_extension_type>&)
+      -> std::shared_ptr<arrow::Array> {
+      VAST_ASSERT(ba->byte_width() == 17);
+      auto subnet_array_data = convert_subnet_array(ba);
+      // TODO: the type is not actually correct, we have
+      // left:  struct<address: fixed_size_binary[16], length: uint8>
+      // right: struct<address: extension<vast.address>, length: uint8>
+      // so we can't create the subnet_array either (arrow asserts storage
+      // type) this gets "fixed" by arrow when re-reading the record batch
+      // in, but the record batch under construction exposes the wrong type
+      // information. to fix this, we require proper AddressBuilder /
+      // SubnetBuilder IIUC
+      // return std::make_shared<subnet_array>(st, subnet_array_data);
+      // this sorta works:  we need to disable verify_record_batch, though.
+      return subnet_array_data;
+    },
+    [&](const std::shared_ptr<arrow::UInt64Array>& a,
+        const std::shared_ptr<enum_extension_type>& et)
+      -> std::shared_ptr<arrow::Array> {
+      auto dict = make_dict_array(et->get_enum_type());
+      auto vals = arrow::compute::Cast(*a, arrow::int16());
+      auto dict_array = std::make_shared<arrow::DictionaryArray>(
+        arrow::dictionary(arrow::int16(), arrow::utf8()), vals.ValueUnsafe(),
+        dict);
+      return std::make_shared<enum_array>(et, dict_array);
+    },
+    [&](const std::shared_ptr<arrow::ListArray>& la,
+        const std::shared_ptr<arrow::MapType>& mt)
+      -> std::shared_ptr<arrow::Array> {
+      return std::make_shared<arrow::MapArray>(mt, la->length(),
+                                               la->value_offsets(),
+                                               la->values(), la->null_bitmap(),
+                                               la->null_count());
+    },
+    [&](const std::shared_ptr<arrow::Array>&,
+        const std::shared_ptr<arrow::StructType>& st)
+      -> std::shared_ptr<arrow::Array> {
+      // move back the iterator so it points to the first array in the struct
+      --array_iterator;
```

Comment:
This reads like a hack to me, but I'm not sure what a better way to implement this would be.

Reply:
It is a hack, indeed. Initially I wanted to avoid `arrow_iterator++` in all branches. But when adding handlers for nests and list types, suddenly `--` is becoming more widespread as well, so I'll make the `++` explicit in according branches.